Add ability for crossgen2 to synthesize PGO histograms

jakobbotsch opened the pull request:
Add an option --synthesize-random-mibc. When passed, crossgen2 will use
metadata from the compilation group to synthesize PGO histograms for
methods that did not have any input profile data. Also supported for --embed-pgo-data,
in which case the synthesized PGO data gets embedded.

Adds a `synthesizepgo` argument to the `src/tests/run.cmd`/`src/tests/run.sh` scripts which enable this mode for compatible tests. So testing can be done locally via:
```
.\src\tests\build.cmd crossgen2 checked
.\src\tests\run.cmd checked runcrossgen2tests synthesizepgo
```

This passes in pri1 on win-x64 for me currently. I have not yet hooked it up in CI. I want to do that in a separate change later, for now this is enough to allow me to do some manual testing on #77267 that has at least some level of coverage.

Some stats:
* crossgen'd SPC, static PGO: 1170 type-based GDVs
* crossgen'd SPC, synthesis: 8400 type-based GDVs
* crossgen'd libraries, static PGO: 2362 type-based GDVs
* crossgen'd libraries, synthesis: 33061 type-based GDVs
* Priority 1 tests with synthesized PGO: 19450 type-based GDVs

Ideally we can get the libraries tests running with crossgen'd libraries using synthesized PGO, but it will depend on #75230, so I will leave it to future work.

jakobbotsch committed:
Add ability for crossgen2 to synthesize PGO histograms

Add an option --synthesize-random-mibc. When passed, crossgen2 will use
metadata from the compilation group to synthesize PGO histograms for
methods that did not have any input profile data.

Mainly for testing purposes. Currently only works in R2R mode, but
hopefully will also work to embed random static PGO data in the future.

jakobbotsch committed:
Fix embedding synthesized PGO data

jakobbotsch committed:
JIT: Support profiles with only histograms

Split up "we have a profile" and "we have profile weights" questions.

jakobbotsch committed:
JitDisasm header nit

jakobbotsch committed:
Fix isSynthesized out var

jakobbotsch committed:
Iterate profile right after retrieving it

Avoid ordering issues. Also revert now unrelated PhasedVar changes.

jakobbotsch commented on src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs:
```diff
@@ -4001,7 +3996,16 @@ public static void ComputeJitPgoInstrumentationSchema(Func<object, IntPtr> objec
+
+                if (pgoResultsSchemas != null && isSynthesized && _compilation.ProfileData.EmbedPgoDataInR2RImage)
+                {
+                    _methodCodeNode.AddSynthesizedPgoDataDependency(pgoResultsSchemas);
```
@davidwrighton The way this currently works is that the `GetAllowSynthesis` call above will both synthesize the schema and "register" it to be physically embedded once `InstrumentationDataTableNode.GetData` is called. This `AddSynthesizedPgoDataDependency` will then cause `MethodWithGCInfo` to return the schema's dependencies from `MethodWithGCInfo.ComputeNonRelocationBasedDependencies`.

It seems to work, but I wonder if the approach is problematic in case the compilation here fails. Do I need to instead register the schema for physical embedding from inside `MethodWithGCInfo.PublishCode` to make sure the physical embedding and `MethodWithGCInfo.ComputeNonRelocationBasedDependencies` are kept "in sync"?

davidwrighton replied:
We may run compilation for a single method more than once. Thus the best practice is to not affect any other part of the system untile the PublishCode logic is run.

jakobbotsch committed:
Revert profile weight schema element

jakobbotsch committed:
Support type based GDV too

jakobbotsch committed:
Avoid empty histograms

jakobbotsch committed:
Merge branch 'main' of github.com:dotnet/runtime into crossgen2-synthesized-pgo

jakobbotsch committed:
Move to PublishCode

jakobbotsch committed:
Remove dead code

jakobbotsch committed:
Nits

jakobbotsch committed:
Remove more dead code, use List<T>.Contains

jakobbotsch committed:
Avoid multiple iteration

jakobbotsch committed:
Fix build

jakobbotsch committed:
Add to crossgen2 test wrapper script invocation

jakobbotsch committed:
Unrelated R2R comment fix

jakobbotsch committed:
Fix concurrency

jakobbotsch committed:
Add some EH

jakobbotsch committed:
Fix wrong metadata name in test

jakobbotsch committed:
Unify RunCrossGen2 environment variable to be set to 1 instead of "true"

Various places were setting this to "true" and various other places to
"1". Always use "1" in the environment variable and "true" in the
msbuild property. This fixes local runs of crossgen2 tests (using
src/tests/run.cmd runcrossgen2tests) -- these were not skipping
FileCheck tests before as the test wrapper script only checks for "1".

jakobbotsch committed:
Also do it for LargeVersionBubble and fix comment

jakobbotsch committed:
Merge branch 'main' of github.com:dotnet/runtime into crossgen2-synthesized-pgo

jakobbotsch committed:
Try to ensure synthesized types/methods are loadable

jakobbotsch committed:
Hook up to testing, mark a test as incompatible with synthesis

jakobbotsch committed:
Fix rerunning crossgen2 tests

jakobbotsch committed:
Merge branch 'main' of github.com:dotnet/runtime into crossgen2-synthesized-pgo

jakobbotsch committed:
Skip generic types/methods in histograms

jakobbotsch committed:
Revert "Unrelated R2R comment fix"

This reverts commit a8ff57f90c39213869d0ceccc0b87a68cbc85ed2.

jakobbotsch committed:
Nit

jakobbotsch commented on src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/ReadyToRunCodegenCompilationBuilder.cs:
```diff
@@ -61,9 +61,6 @@ public sealed class ReadyToRunCodegenCompilationBuilder : CompilationBuilder
             _ilProvider = new ReadyToRunILProvider(group);
             _inputFiles = inputFiles;
             _compositeRootPath = compositeRootPath;
-
-            // R2R field layout needs compilation group information
-            ((ReadyToRunCompilerContext)context).SetCompilationGroup(group);
```
I've moved this so it happens right after we create the compilation group so it is early enough for the `ProfileDataManager` that also needs this now (for some `EnsureLoadable` calls).

jakobbotsch committed:
Nit

jakobbotsch committed:
Check IsTypicalMethodDefinition for delegate invoke too

jakobbotsch commented on src/tests/Loader/classloader/MethodImpl/CovariantReturns/UnitTest/CompatibleWithTest.ilproj:
```diff
@@ -1,6 +1,8 @@
 <Project Sdk="Microsoft.NET.Sdk.IL">
   <PropertyGroup>
     <OutputType>Exe</OutputType>
+    <!-- Synthesized PGO may cause the expected TypeLoadExceptions in this test to appear at unexpected places -->
+    <SynthesizedPgoIncompatible>true</SynthesizedPgoIncompatible>
```
This is probably an actual bug -- loading the type for the GDV guess can have observable behavior on load failure, and does so in this test. Ideally we would have a "TypeHandle or null if not loadable" fixup we could use for GDV guesses, but for now I have just disabled this single test.

jakobbotsch committed:
Add some comments

jakobbotsch commented:
cc @dotnet/jit-contrib PTAL @davidwrighton

jakobbotsch committed:
Merge branch 'main' of github.com:dotnet/runtime into crossgen2-synthesized-pgo

jakobbotsch committed:
Unify crossgen2 and ILC's additional dependencies handling

davidwrighton reviewed (approved):
:shipit:

jakobbotsch commented:
/azp run runtime-coreclr outerloop

jakobbotsch committed:
Merge branch 'main' of github.com:dotnet/runtime into crossgen2-synthesized-pgo

jakobbotsch committed:
Fix build after merge
